src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -299,6 +307,115 @@ static void AppendHashtableContents(IndentedTextWriter writer, Hashtable ht)
             }
         }
 
+        private static void EmitScan(IndentedTextWriter writer, RegexMethod rm, string id)
+        {
+            RegexOptions options = (RegexOptions)rm.Options;
+
+            // Emit locals initialization
+            //writer.WriteLine("this.quick = quick;");
+            //writer.WriteLine("base.runtextpos = textstart;");
+            //writer.WriteLine("base.runregex = regex;");
+            //writer.WriteLine("base.runtextstart = textstart;");
+            //writer.WriteLine("base.runtextbeg = textbeg;");
+            //writer.WriteLine("base.runtextend = textend;");
+            //writer.WriteLine("bool initialized = false;");
+            //writer.WriteLine();
+
+            //EmitTimeoutHandling();
+            //writer.WriteLine();
+
+            // Source generator doesn't support Right-To-Left so there is no need to add the sepcial bump logic.
+            // If Right-to-left is ever added to the source generator, then we would need to the logic to define
+            // bump, as well as stoppos
+            Debug.Assert((options & RegexOptions.RightToLeft) == 0);
+
+            EmitPrevLenCheck();
+            writer.WriteLine();
+
+            EmitMainSearchLoop();
+            writer.WriteLine();
+
+            return;
+
+#pragma warning disable CS8321 // Local function is declared but never used
+            void EmitTimeoutHandling()
+            {
+                writer.WriteLine("// Handle timeout argument");
+                writer.WriteLine("_timeout = -1;");
+                writer.WriteLine("bool ignoreTimeout = global::System.Text.RegularExpressions.Regex.InfiniteMatchTimeout == timeout;");
+                using (EmitBlock(writer, "if (!ignoreTimeout)"))
+                {
+                    writer.WriteLine("// We are using Environment.TickCount and not Stopwatch for performance reasons.");
+                    writer.WriteLine("// Environment.TickCount is an int that cycles. We intentionally let timeoutOccursAt");
+                    writer.WriteLine("// overflow it will still stay ahead of Environment.TickCount for comparisons made");
+                    writer.WriteLine("// in DoCheckTimeout().");
+                    writer.WriteLine("global::System.Text.RegularExpressions.Regex.ValidateMatchTimeout(timeout); // validate timeout as this could be called from user code due to being protected");
```

Comment:
I wonder if we really need this.  Presumably either the timeout is getting passed into this from something in the library, in which case it can already have validated it, or it's coming from something in the source-generated implementation, at which point all of this code is the user's and if they want to mess things up by supplying a nonsense timeout, they're only messing up their own code.

I could be wrong about that, but it'd be worth double-checking.

Reply:
The idea is that overrides of Scan should be calling already protected method `CheckTimeout` when relevant. Timeout will already be setup beforehand correctly, and CheckTimeout will only perform a quick if-check in case no timeout was provided.